sys/usb/usbus/dfu/dfu.c

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (C) 2020 Dylan Laduranty
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup usbus_dfu
+ * @{
+ * @file USBUS implementation for device firmware upgrade
+ * 
+ * @note This feature is currently experimental as it needs a proper riotboot
+ * or another bootloader implementation to be fully functional
+ * 
+ *
+ * @author  Dylan Laduranty <dylan.laduranty@mesotic.com>
+ * @}
+ */
+
+#define USB_H_USER_IS_RIOT_INTERNAL
+
+#include "usb/dfu.h"
+#include "usb/descriptor.h"
+#include "usb/usbus.h"
+#include "usb/usbus/control.h"
+#include "xtimer.h"
+
+#include "periph/flashpage.h"
+#include "periph/pm.h"
+
+#include <string.h>
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static void _event_handler(usbus_t *usbus, usbus_handler_t *handler,
+                          usbus_event_usb_t event);
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                            usbus_control_request_state_t state,
+                            usb_setup_t *setup);
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                              usbdev_ep_t *ep, usbus_event_transfer_t event);
+static void _init(usbus_t *usbus, usbus_handler_t *handler);
+
+static usbus_dfu_state_t dfu_state;
+static uint32_t offset = 0;
+static uint8_t page_mem[FLASHPAGE_SIZE];
+/* TODO: properly integrate this part to riotboot */
+static usbus_dfu_device_t dfu2;
+static char _stack[USBUS_STACKSIZE];
+static usbus_t usbus2;
+
+static size_t _gen_if_descriptor(usbus_t *usbus)
+{
+    usb_desc_if_dfu_t if_desc;
+
+    /* functional dfu descriptor */
+    if_desc.length = sizeof(usb_desc_if_dfu_t);
+    if_desc.type = USB_IF_DESCRIPTOR_DFU;
+    if_desc.attribut = USB_DFU_WILL_DETACH | USB_DFU_CAN_DOWNLOAD;
+    if_desc.detach_timeout = 255;
+    if_desc.xfer_size = FLASHPAGE_SIZE;
```

Comment:
These two should be defines I think:
```suggestion
    if_desc.detach_timeout = USB_DFU_DETACH_TIMEOUT;
    if_desc.xfer_size = USB_DFU_TRANSFER_SIZE;
```

`USB_DFU_TRANSFER_SIZE` can be defined as `FLASHPAGE_SIZE`

Reply:
Done.